clients/drcachesim/tests/burst_traceopts.cpp

```diff
@@ -0,0 +1,363 @@
+/* **********************************************************
+ * Copyright (c) 2019 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Tests offline trace recording optimizations and ensures that elided information
+ * is accurately reconstructed in post-processing by collecting two traces of an
+ * identical code region, one with and one without optimizations.  These two
+ * traces are post-processed and compared, all within this test.
+ */
+
+#ifndef ASM_CODE_ONLY /* C code */
+#    include "dr_api.h"
+#    include "drmemtrace/drmemtrace.h"
+#    include "drcovlib.h"
+#    include "analysis_tool.h"
+#    include "analyzer.h"
+#    include "tracer/raw2trace.h"
+#    include "tracer/raw2trace_directory.h"
+#    include <assert.h>
+#    include <iostream>
+#    include <math.h>
+#    include <stdlib.h>
+#    include <string.h>
+
+/* Asm routines. */
+extern "C" {
+void
+test_disp_elision();
+void
+test_base_elision();
+};
+
+class trace_comparer_t : public analysis_tool_t {
```

Comment:
Hmm.  On the one hand it is logically a trace analysis tool, which is why I wrote it like one, but OTOH this exposed iterator lets us just directly call anything we want, so you're right that it can just be a function.  It does have some state, with a counter for better diagnostics on mismatches, but that can be solved.  Sure, will do.